tests/test_tomography_gate_set_tomography.py

```diff
@@ -0,0 +1,2535 @@
+import numpy as np
+import pytest
+
+import qibo
+from qibo import gates
+from qibo.noise import DepolarizingError, NoiseModel
+from qibo.tomography.gate_set_tomography import (
+    GST_execute_circuit,
+    execute_GST,
+    measurement_basis,
+    prepare_states,
+    reset_register,
+)
+
+
+def test_prepare_states_valid_k_single_qubit():
+    # Test for valid input with a single qubit
+    k = 2
+    nqubits = 1
+
+    circuit = prepare_states(k, nqubits)
+
+    assert isinstance(circuit, qibo.models.circuit.Circuit)
+    assert circuit.nqubits == nqubits
+
+
+def test_prepare_states_valid_k_two_qubits():
+    # Test for valid input with two qubits
+    k = 1
+    nqubits = 2
+
+    circuit = prepare_states(k, nqubits)
+
+    assert isinstance(circuit, qibo.models.circuit.Circuit)
+    assert circuit.nqubits == nqubits
+
+
+def test_prepare_states_valid_k_invalid_nqubits():
+    # Test for value input with invalid nqubits
+    k = 0
+    nqubits = 3
+
+    # Check if ValueError is raised
+    with pytest.raises(ValueError, match="nqubits needs to be either 1 or 2"):
+        circuit = prepare_states(k, nqubits)
+
+
+def test_prepare_states_invalid_k_valid_nqubits():
+    # Test for invalid input with valid nqubits
+    k = 17
+    nqubits = 2
+
+    # Check if IndexError is raised
+    with pytest.raises(IndexError):
+        circuit = prepare_states(k, nqubits)
```

Comment:
All these tests can be reduced to a single one using pytest parameters, have a look here [https://docs.pytest.org/en/7.3.x/how-to/parametrize.html](https://docs.pytest.org/en/7.3.x/how-to/parametrize.html).
The same applies for all the following tests.

Reply:
Let me add one more thing: one could look at other test files to see how `pytest.mark.parametrize` works and also how the tests should work for all backends